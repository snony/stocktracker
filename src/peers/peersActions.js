export const PEERS_RECEIVED_ACTION = 'PEERS_RECEIVED_ACTION'

const peersReceivedAction = peers => ({ type: PEERS_RECEIVED_ACTION, peers })

export const getPeersData = symbol => (dispatch, _, api) => {
  api.getPeers(symbol).then(peers => {
    return dispatch(peersReceivedAction(peers))
  })
}
