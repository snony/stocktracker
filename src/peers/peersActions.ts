import { Action, ActionCreator } from 'redux'
import { ThunkAction } from 'redux-thunk'

import { API, GlobalState } from '../types'
import { Peers } from './types'

export const PEERS_RECEIVED_ACTION = 'PEERS_RECEIVED_ACTION'

export interface PeersReceivedAction extends Action {
  type: typeof PEERS_RECEIVED_ACTION
  peers: Peers
}

const peersReceivedAction: ActionCreator<PeersReceivedAction> = (peers: string[]) => ({
  type: PEERS_RECEIVED_ACTION,
  peers
})

type ThunkResult<R> = (symbol: string) => ThunkAction<R, GlobalState, API, PeersReceivedAction>

export const getPeersData: ThunkResult<void> = (symbol: string) => (dispatch, _, api) => {
  api.getPeers(symbol).then(peers => dispatch(peersReceivedAction(peers)))
}
