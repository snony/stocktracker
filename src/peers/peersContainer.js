import { connect } from 'react-redux'
import Peers from './peers'

const mapStateToProps = state => ({
  peers: state.companyInfo.peers
})

export default connect(mapStateToProps)(Peers)
