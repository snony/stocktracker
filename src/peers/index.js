import PeersContainer from './peersContainer'
import { getPeersData } from './peersActions'
import peers from './peersReducer'

export { getPeersData, peers }
export default PeersContainer
