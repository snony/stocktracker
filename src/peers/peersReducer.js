import { PEERS_RECEIVED_ACTION } from './peersActions'

const initialState = []

export default (state = initialState, { type, peers }) => {
  switch (type) {
    case PEERS_RECEIVED_ACTION:
      return [...peers]
    default:
      return state
  }
}
