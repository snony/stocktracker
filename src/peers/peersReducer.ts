import { Reducer } from 'redux'

import { PEERS_ACTION_TYPES, PeersActions } from './peersActions'
import { Peers } from './types'

export interface PeersState {
  peers: Peers,
  error: boolean
}

export const initialState: PeersState = {
  peers: [],
  error: false
}

const peersReducer: Reducer<PeersState, PeersActions> = (state = initialState, action) => {
  switch (action.type) {
    case PEERS_ACTION_TYPES.RECEIVED_DATA:
      return {
        ...state,
        peers: action.peers,
        error: false
      }
    case PEERS_ACTION_TYPES.RECEIVED_ERROR:
      return {
        ...state,
        error: true
      }
    default:
      return state
  }
}

export default peersReducer
