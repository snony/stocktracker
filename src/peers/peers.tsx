import React from 'react'

import { peersContainerState } from './peersContainer'

type PeersProps = peersContainerState

const Peers: React.SFC<PeersProps> = ({ peers }) => {
  if(peers.length !== 0) {
    const peersData = peers.join(', ')
    return <span className="label label--small">{peersData}</span>
  } else {
    return <p className="label label--small">Loading...</p>
  }
}

export default Peers
