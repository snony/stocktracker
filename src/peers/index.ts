import { getPeersData } from './peersActions'
import PeersContainer from './peersContainer'
import peers from './peersReducer'

export { getPeersData, peers }
export default PeersContainer
