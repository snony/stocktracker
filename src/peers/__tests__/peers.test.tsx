import { mockFailedGlobalState, mockGlobalState } from '__mock__/globalstate.mock'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import Peers from 'peers/peers'
import { initialState } from 'peers/peersReducer'
import React from 'react'

describe('Peers Component', () => {
  it('should render Peers with loading... when peers state is empty', () => {
    const wrapper = shallow(<Peers peers={initialState} />)

    expect(toJSON(wrapper)).toMatchSnapshot()
  })

  it('should render Peers with peers state data correctly', () => {
    const wrapper = shallow(<Peers peers={mockGlobalState.peers} />)

    expect(toJSON(wrapper)).toMatchSnapshot()
  })

  it('should render Peers with Error when peers state is error', () => {
    const wrapper = shallow(<Peers peers={mockFailedGlobalState.peers} />)

    expect(toJSON(wrapper)).toMatchSnapshot()
  })
})
