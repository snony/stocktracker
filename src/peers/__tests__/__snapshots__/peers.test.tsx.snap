// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Peers Component should render Peers with Error when peers state is error 1`] = `
<Styled(span)
  grey={true}
  small={true}
>
  Loading...
</Styled(span)>
`;

exports[`Peers Component should render Peers with loading... when peers state is empty 1`] = `
<Styled(span)
  grey={true}
  small={true}
>
  Loading...
</Styled(span)>
`;

exports[`Peers Component should render Peers with peers state data correctly 1`] = `
<Styled(span)
  grey={true}
  small={true}
>
  peers1, peers2, peers3, peers4
</Styled(span)>
`;
